### PR TITLE
Fix the search, similar, list, and developer methods

### PR DESCRIPTION
**Pull request by mmoksh**
Google introduced new mapping for the first page of the search results to highlight one app that matches the search query. The rest of the pages have the same mapping as previous.

In this PR, I fixed `gplay.search` script by updating mappings. I also included the changes from PR #545 that fixes `gplay.app` in this PR to avoid git conflicts when merging both PRs.

Edit: Added a few more commits to fix the `gplay.similar` script, and to handle special cases when there is no search results.

Edit2: Added a new commit to fix the `gplay.list` method.

Use this for testing:
```
"google-play-scraper": "git+https://github.com/facundoolano/google-play-scraper#pull/557/head"
```

### CODE REVIEW

**Comment by Aliaksandr-Kasko-JazzTeam** [comment]
Because of changes to GPlay you should update tests to fix failing checks.

**Comment by Aliaksandr-Kasko-JazzTeam** [comment]
There's no fails with list method, but it returns an empty array for all combinations of category and collection :-\

**Comment by NathanManning** [comment]
@mmoksh, did something change? I'm getting empty results again for the clustered apps.

`gplay
    .search({
        term: 'facebook',
    })
    .then(console.log, console.log)`
    
    returns []

**Comment by mmoksh** [comment]
@NathanManning I just checked! It works just fine!
@Aliaksandr-Kasko-JazzTeam This PR does not contain a fix for the `gplay.list` method. It only fixes the `gplay.search` and `gplay.similar` methods. I'll try to update the unit tests later this week.

**Comment by Aliaksandr-Kasko-JazzTeam** [comment]
@mmoksh could you please fix `gplay.list` method too? It's really necessary to us

**Comment by NathanManning** [comment]
> @NathanManning I just checked! It works just fine! @Aliaksandr-Kasko-JazzTeam This PR does not contain a fix for the `gplay.list` method. It only fixes the `gplay.search` and `gplay.similar` methods. I'll try to update the unit tests later this week.

@mmoksh  My bad. There's something funky going on in our package-lock.

**Comment by mmoksh** [comment]
@Aliaksandr-Kasko-JazzTeam The `gplay.list` method is fixed now. You might need to delete the `package-lock.json` file and the `node_modules` folder to make sure your app is not using an outdated version of this PR.

**Comment by ninjachen** [comment]
Hello, I'm trying to use the new list function of this PR and meet an error, could you help

The steps to reproduce
```
➜  scraper git:(master) ✗ npm -v
7.24.0
➜  scraper git:(master) ✗ rm -rf node_modules 
➜  scraper git:(master) ✗ rm package-lock.json 
➜  scraper git:(master) ✗ npm install
➜  scraper git:(master) ✗ npx functions-framework --target=list
node:internal/modules/cjs/loader:926
  throw err;
  ^

Error: Cannot find module './middleware/cloudevent_to_background_event'
Require stack:
- /Users/ninja/src/lab/scraper/node_modules/@google-cloud/functions-framework/build/src/server.js
- /Users/ninja/src/lab/scraper/node_modules/@google-cloud/functions-framework/build/src/index.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:923:15)
    at Function.Module._load (node:internal/modules/cjs/loader:768:27)
    at Module.require (node:internal/modules/cjs/loader:995:19)
    at require (node:internal/modules/cjs/helpers:92:18)
    at Object.<anonymous> (/Users/ninja/src/lab/scraper/node_modules/@google-cloud/functions-framework/build/src/server.js:24:42)
    at Module._compile (node:internal/modules/cjs/loader:1091:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1120:10)
    at Module.load (node:internal/modules/cjs/loader:971:32)
    at Function.Module._load (node:internal/modules/cjs/loader:812:14)
    at Module.require (node:internal/modules/cjs/loader:995:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/Users/ninja/src/lab/scraper/node_modules/@google-cloud/functions-framework/build/src/server.js',
    '/Users/ninja/src/lab/scraper/node_modules/@google-cloud/functions-framework/build/src/index.js'
  ]
}
```

package.json
```
➜  scraper git:(master) ✗ cat package.json 
{
  "name": "taptap-scraper",
  "version": "0.5.1",
  "dependencies": {
    "google-play-scraper": "git+https://github.com/facundoolano/google-play-scraper#pull/557/head"
  },
  "scripts": {
    "start": "npx functions-framework --target=app --signature-type=http"
  },
  "devDependencies": {
    "@google-cloud/functions-framework": "^1.9.0"
  }
}
```

**Comment by Aliaksandr-Kasko-JazzTeam** [comment]
@mmoksh many thanks to you. It seems to work fine. 👍 

@facundoolano This PR looks great to merge into master and new package version

**Comment by mmoksh** [comment]
> Hello, I'm trying to use the new list function of this PR and meet an error, could you help
> 
> The steps to reproduce
> 
> ```
> ➜  scraper git:(master) ✗ npm -v
> 7.24.0
> ➜  scraper git:(master) ✗ rm -rf node_modules 
> ➜  scraper git:(master) ✗ rm package-lock.json 
> ➜  scraper git:(master) ✗ npm install
> ➜  scraper git:(master) ✗ npx functions-framework --target=list
> 
> Serving function...
> Function: list
> Signature type: http
> URL: http://localhost:8080/
> Error: Invalid collection topselling_free
>     at validate (/Users/ninja/src/lab/scraper/node_modules/google-play-scraper/lib/list.js:43:11)
>     at /Users/ninja/src/lab/scraper/node_modules/google-play-scraper/lib/list.js:14:5
>     at new Promise (<anonymous>)
>     at Object.list (/Users/ninja/src/lab/scraper/node_modules/google-play-scraper/lib/list.js:13:10)
>     at exports.list (/Users/ninja/src/lab/scraper/index.js:31:15)
>     at /Users/ninja/src/lab/scraper/node_modules/@google-cloud/functions-framework/build/src/invoker.js:98:17
>     at processTicksAndRejections (node:internal/process/task_queues:76:11)
> ```
> 
> package.json
> 
> ```
> ➜  scraper git:(master) ✗ cat package.json 
> {
>   "name": "scraper",
>   "version": "0.5.1",
>   "dependencies": {
>     "google-play-scraper": "git+https://github.com/facundoolano/google-play-scraper#pull/557/head"
>   },
>   "scripts": {
>     "start": "npx functions-framework --target=app --signature-type=http"
>   },
>   "devDependencies": {
>     "@google-cloud/functions-framework": "1.9.0"
>   }
> }
> ```

I think `topselling_free` collection is not available on Google Play anymore. I could see `Top free`, `Top grossing`, and `Top paid` on the website, but not top selling. I updated the tests to avoid this error.

File: lib/app.js [hunk 1]
```diff
@@ -43,169 +42,113 @@ function app (opts) {
...
+    fun: (val) => val || 'VARY'
   },
-  editorsChoice: {
+  recentChanges: ['ds:4', 1, 2, 144, 1, 1],
+  comments: {
+    path: ['ds:8', 0],
+    isArray: true,
+    fun: helper.extractComments
+  }
+/*   editorsChoice: {
```
Comment (facundoolano) [inline lib/app.js:144]: is this one missing in the new page? if so we should just remove it

File: test/lib.categories.js [hunk 2]
```diff
@@ -20,11 +20,12 @@ describe('Categories method', () => {
         [],
         'Google Play has categories that are not in "category" constant'
       );
-      assert.deepEqual(
-        R.difference(categoriesConst, categories),
-        [],
-        'There are some categories in the constant that were removed from Play Store'
-      );
+      // This url is not active anymore: https://play.google.com/store/apps/category
```
Comment (facundoolano) [inline test/lib.categories.js:23]: remove

File: lib/app.js (hunk 1 above)
Comment (mmoksh) [reply lib/app.js]: Just removed it.

File: lib/utils/appList.js [hunk 3]
```diff
@@ -53,4 +50,4 @@ function extract (root, data) {
   return R.map(scriptData.extractor(MAPPINGS), input);
 }
 
-module.exports = { MAPPINGS, extract };
+module.exports = { MAPPINGS, extract, extaractDeveloperId };
```
[the remark this answers is not in the document]
Comment (mmoksh) [reply lib/utils/appList.js]: Done.

File: test/lib.app.js [hunk 4]
```diff
@@ -19,7 +19,7 @@ const validateAppDetails = (app) => {
   assert.isString(app.summary);
   assert.isString(app.description);
   assert.isString(app.descriptionHTML);
-  assert.isNumber(app.updated);
+  // assert.isNumber(app.updated);
```
[the remark this answers is not in the document]
Comment (mmoksh) [reply test/lib.app.js]: Done.

**Comment by mmoksh** [comment]
> @mmoksh thanks for this.
> 
> This one PR addresses all the methods right? I'll set aside some time to review it, merge and publish a new version of the package. Please let me know if there something missing or broken still that I should take into account (I can't keep up with all the conversations going on in issues and PRs)

Sure!
The PR covers all the methods.

File: lib/list.js [hunk 5]
```diff
@@ -48,114 +65,62 @@ function validate (opts) {
...
   const queryString = {
-    hl: 'us',
+    hl: opts.lang,
     gl: opts.country
   };
 
   if (opts.age) {
     queryString.age = opts.age;
   }
+  const url = `https://play.google.com/_/PlayStoreUi/data/batchexecute?rpcids=vyAe2&source-path=%2Fstore%2Fapps&f.sid=-4178618388443751758&bl=boq_playuiserver_20220612.08_p0&authuser=0&soc-app=121&soc-platform=1&soc-device=1&_reqid=82003&rt=c`;
```
Comment (nadavzipo) [inline lib/list.js:77]: @mmoksh should use ${BASE_URL} here as well.

@facundoolano will you approve a PR where the baseURL can be passed as a parameter, so we can use services like scraperstack?

Comment (facundoolano) [reply lib/list.js:77]: +1 to using BASE_URL here

@nadavzipo I would approve but that should be done in a separate PR after this one gets merged

File: lib/app.js [hunk 6]
```diff
@@ -43,169 +42,105 @@ function app (opts) {
...
   },
   offersIAP: {
-    path: ['ds:5', 0, 12, 12, 0],
+    path: ['ds:4', 1, 2, 19, 0],
     fun: Boolean
   },
-  IAPRange: ['ds:5', 0, 12, 12, 0],
-  size: ['ds:8', 0],
+  IAPRange: ['ds:4', 1, 2, 19, 0],
+  /* size: ['ds:8', 0], */
```
Comment (facundoolano) [inline lib/app.js:86]:
```suggestion
```

File: lib/list.js (hunk 5 above)
Comment (facundoolano) [reply lib/list.js:77]:
```suggestion
  const url = `${BASE_URL}/_/PlayStoreUi/data/batchexecute?rpcids=vyAe2&source-path=%2Fstore%2Fapps&f.sid=-4178618388443751758&bl=boq_playuiserver_20220612.08_p0&authuser=0&soc-app=121&soc-platform=1&soc-device=1&_reqid=82003&rt=c`;
```

File: test/lib.app.js [hunk 7]
```diff
@@ -59,9 +58,9 @@ const validateAppDetails = (app) => {
 
   assert.isString(app.recentChanges);
 
-  assert.isBoolean(app.editorsChoice);
+  // assert.isBoolean(app.editorsChoice);
```
Comment (facundoolano) [inline test/lib.app.js:61]:
```suggestion
```

File: test/lib.app.js [hunk 8]
```diff
@@ -59,9 +58,8 @@ const validateAppDetails = (app) => {
 
   assert.isString(app.recentChanges);
 
-  assert.isBoolean(app.editorsChoice);
 
-  assert.equal(2, app.features.length);
+  // assert.equal(2, app.features.length);
```
Comment (facundoolano) [inline test/lib.app.js:62]:
```suggestion
```

**Pull request merged**